File: cli/test/ship.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { createServer } from 'node:http'
import { mkdirSync, mkdtempSync, readFileSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { ConfigError } from '../src/config.js'
import { parseDeploy, releaseId, uniqueRelease, RELEASE_ID } from '../src/ship/config.js'
import { runShip } from '../src/ship/index.js'

const base = { path: '/srv/app', upload: ['dist'] }

test('минимальная секция выкатки разбирается с разумными умолчаниями', () => {
  const d = parseDeploy(base, 'deploy')
  assert.equal(d.host, null)
  assert.equal(d.keep, 5)
  assert.equal(d.checks, true)
  assert.equal(d.health, null)
})

test('здоровье задаётся строкой или объектом', () => {
  assert.deepEqual(parseDeploy({ ...base, health: 'https://x/health' }, 'deploy').health, {
    url: 'https://x/health',
    timeout: 60,
  })
  assert.deepEqual(parseDeploy({ ...base, health: { url: 'http://x', status: 204, timeout: 5 } }, 'deploy').health, {
    url: 'http://x',
    status: 204,
    timeout: 5,
  })
})

test('без path и без upload выкатка не описана', () => {
  assert.throws(() => parseDeploy({ upload: ['dist'] }, 'deploy'), ConfigError)
  assert.throws(() => parseDeploy({ path: '/srv/app' }, 'deploy'), ConfigError)
  assert.throws(() => parseDeploy({ ...base, upload: [] }, 'deploy'), ConfigError)
})

test('пути отправки не выпускаются за пределы проекта', () => {
  for (const bad of ['/etc/passwd', '../секреты', 'C:\\Windows', 'dist/../../..']) {
    assert.throws(
      () => parseDeploy({ ...base, upload: [bad] }, 'deploy'),
      ConfigError,
      `путь "${bad}" должен быть отвергнут`,
    )
  }
})

test('keep меньше двух отвергается — откатываться было бы некуда', () => {
  assert.throws(() => parseDeploy({ ...base, keep: 1 }, 'deploy'), ConfigError)
  assert.equal(parseDeploy({ ...base, keep: 2 }, 'deploy').keep, 2)
})

test('имя выпуска сортируется по алфавиту так же, как по времени', () => {
  const earlier = releaseId(new Date('2026-09-07T17:22:56Z'))
  const later = releaseId(new Date('2026-09-07T17:22:57Z'))
  assert.equal(earlier, '20260907-172256')
  assert.ok(earlier < later)
  assert.match(earlier, RELEASE_ID)
})

test('две выкатки в одну секунду получают разные имена', () => {
  const first = uniqueRelease([], '20260907-172256')
  const second = uniqueRelease([first], '20260907-172256')
  const third = uniqueRelease([first, second], '20260907-172256')

  assert.notEqual(first, second)
  assert.ok(first < second && second < third, 'порядок имён должен сохраняться')
  for (const id of [first, second, third]) assert.match(id, RELEASE_ID)
})

/**
 * Полный цикл выкатки на локальной цели.
 *
 * Проверяется главное обещание команды: испорченный выпуск не остаётся
 * работать. Приложение здесь — файл version.txt, который читает поднятый в
 * тесте сервер; он отвечает 500, если содержимое не то. Так проверка здоровья
 * настоящая, а не имитация.
 */
test('негодный выпуск откатывается, и продолжает работать прежний', async () => {
  const project = mkdtempSync(join(tmpdir(), 'p3k-ship-src-'))
  const server = mkdtempSync(join(tmpdir(), 'p3k-ship-dst-'))
  mkdirSync(join(project, 'app'))
  writeFileSync(join(project, 'package.json'), '{"name":"ship-demo","private":true}')

  const http = createServer((_req, res) => {
    let body = ''
    try {
      body = readFileSync(join(server, 'current', 'app', 'version.txt'), 'utf8').trim()
    } catch {
      body = 'нет файла'
    }
    res.writeHead(body === 'годная' ? 200 : 500).end(body)
  })
  await new Promise<void>((r) => http.listen(0, '127.0.0.1', r))
  const port = (http.address() as { port: number }).port

  writeFileSync(
    join(project, 'p3k.json'),
    JSON.stringify({
      processes: { app: { command: 'node app.mjs' } },
      deploy: {
        path: server,
        upload: ['app'],
        checks: false,
        health: { url: `http://127.0.0.1:${port}/`, timeout: 4 },
      },
    }),
  )

  // Вывод команды забираем себе: в отчёте теста он лишний, а проверять по нему
  // всё равно нужно.
  const cwd = process.cwd()
  let log = ''
  const io = { out: (s: string) => (log += s), err: (s: string) => (log += s) }
  process.chdir(project)

  try {
    writeFileSync(join(project, 'app', 'version.txt'), 'годная')
    assert.equal(await runShip([], io), 0, 'здоровый выпуск должен выкатиться')
    assert.equal(readFileSync(join(server, 'current', 'app', 'version.txt'), 'utf8'), 'годная')

    const first = readFileSync(join(server, 'current', 'app', 'version.txt'), 'utf8')
    assert.equal(first, 'годная')

    writeFileSync(join(project, 'app', 'version.txt'), 'сломанная')
    assert.equal(await runShip([], io), 1, 'нездоровый выпуск должен провалиться')

    // Главное: цель вернулась к прежнему выпуску, а не осталась сломанной.
    assert.equal(
      readFileSync(join(server, 'current', 'app', 'version.txt'), 'utf8'),
      'годная',
      'после отката должен работать прежний выпуск',
    )
    assert.match(log, /откат на \d{8}-\d{6}/)

    const res = await fetch(`http://127.0.0.1:${port}/`)
    assert.equal(res.status, 200, 'сервер снова отвечает')

    // Оба выпуска остаются на цели: сломанный нужен, чтобы посмотреть логи.
    log = ''
    assert.equal(await runShip(['--releases'], io), 0)
    assert.equal((log.match(/\d{8}-\d{6}/g) ?? []).length, 2)
  } finally {
    process.chdir(cwd)
    http.close()
  }
})

test('без секции deploy команда объясняет, чего не хватает', async () => {
  const project = mkdtempSync(join(tmpdir(), 'p3k-ship-none-'))
  writeFileSync(join(project, 'package.json'), '{"name":"нет-выкатки","private":true}')
  writeFileSync(join(project, 'p3k.json'), '{"processes":{"app":{"command":"node ."}}}')

  const cwd = process.cwd()
  let log = ''
  process.chdir(project)

  try {
    assert.equal(await runShip([], { out: () => {}, err: (s) => (log += s) }), 2)
    assert.match(log, /нет секции deploy/)
  } finally {
    process.chdir(cwd)
  }
})

File: go/ship_test.go
package p3k_test

import (
	"context"
	"fmt"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	p3k "github.com/busyburgerr/p3k/go"
)

// Полный цикл выкатки на локальной цели.
//
// Проверяется главное обещание: испорченный выпуск не остаётся работать.
// Приложение здесь — файл version.txt, который читает поднятый в тесте сервер;
// он отвечает 500, если содержимое не то. Так проверка здоровья настоящая, а
// не имитация.
func TestShipRollsBackUnhealthyRelease(t *testing.T) {
	project := t.TempDir()
	server := t.TempDir()
	if err := os.Mkdir(filepath.Join(project, "app"), 0o755); err != nil {
		t.Fatal(err)
	}

	srv := &http.Server{Handler: http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		body, err := os.ReadFile(filepath.Join(server, "current", "app", "version.txt"))
		if err != nil {
			http.Error(w, "нет файла", http.StatusInternalServerError)
			return
		}
		if strings.TrimSpace(string(body)) != "годная" {
			http.Error(w, string(body), http.StatusInternalServerError)
			return
		}
		w.Write(body)
	})}
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	go srv.Serve(ln)
	defer srv.Close()

	config := fmt.Sprintf(`{
	  "deploy": {
	    "path": %q,
	    "upload": ["app"],
	    "checks": false,
	    "health": { "url": "http://%s/", "timeout": 4 }
	  }
	}`, server, ln.Addr())
	if err := os.WriteFile(filepath.Join(project, "p3k.json"), []byte(config), 0o644); err != nil {
		t.Fatal(err)
	}

	cfg, err := p3k.Load(filepath.Join(project, "p3k.json"))
	if err != nil {
		t.Fatal(err)
	}
	version := filepath.Join(project, "app", "version.txt")
	ctx := context.Background()

	// Первая выкатка: здоровая.
	if err := os.WriteFile(version, []byte("годная"), 0o644); err != nil {
		t.Fatal(err)
	}
	first, err := p3k.Ship(ctx, cfg)
	if err != nil {
		t.Fatalf("здоровый выпуск должен выкатиться: %v", err)
	}
	if !first.Healthy || first.Previous != "" {
		t.Errorf("первая выкатка описана неверно: %+v", first)
	}
	live := filepath.Join(server, "current", "app", "version.txt")
	if data, _ := os.ReadFile(live); string(data) != "годная" {
		t.Fatalf("через current читается %q", data)
	}

	// Вторая: сломанная. Секунда паузы нужна, чтобы имя выпуска отличалось
	// временем, а не порядковым номером, — так нагляднее.
	time.Sleep(time.Second)
	if err := os.WriteFile(version, []byte("сломанная"), 0o644); err != nil {
		t.Fatal(err)
	}
	second, err := p3k.Ship(ctx, cfg)
	if err == nil {
		t.Fatal("нездоровый выпуск должен провалиться")
	}
	if second == nil || !second.RolledBack {
		t.Fatalf("отката не было: %+v (%v)", second, err)
	}

	// Главное: цель вернулась к прежнему выпуску, а не осталась сломанной.
	if data, _ := os.ReadFile(live); string(data) != "годная" {
		t.Errorf("после отката через current читается %q, ожидали прежний выпуск", data)
	}
	res, err := http.Get("http://" + ln.Addr().String() + "/")
	if err != nil || res.StatusCode != http.StatusOK {
		t.Errorf("сервер не отвечает после отката: %v", err)
	}

	// Оба выпуска остаются на цели: сломанный нужен, чтобы посмотреть логи.
	list, current, err := p3k.Releases(ctx, cfg)
	if err != nil {
		t.Fatal(err)
	}
	if len(list) != 2 {
		t.Errorf("выпусков на цели %d, ожидали 2: %v", len(list), list)
	}
	if current != first.Release {
		t.Errorf("работает %q, ожидали %q", current, first.Release)
	}
}

func TestRollbackWalksBackOneRelease(t *testing.T) {
	project := t.TempDir()
	server := t.TempDir()
	if err := os.Mkdir(filepath.Join(project, "app"), 0o755); err != nil {
		t.Fatal(err)
	}

	config := fmt.Sprintf(`{"deploy": {"path": %q, "upload": ["app"], "checks": false}}`, server)
	if err := os.WriteFile(filepath.Join(project, "p3k.json"), []byte(config), 0o644); err != nil {
		t.Fatal(err)
	}
	cfg, err := p3k.Load(filepath.Join(project, "p3k.json"))
	if err != nil {
		t.Fatal(err)
	}

	ctx := context.Background()
	version := filepath.Join(project, "app", "version.txt")

	os.WriteFile(version, []byte("первый"), 0o644)
	first, err := p3k.Ship(ctx, cfg)
	if err != nil {
		t.Fatal(err)
	}

	time.Sleep(time.Second)
	os.WriteFile(version, []byte("второй"), 0o644)
	second, err := p3k.Ship(ctx, cfg)
	if err != nil {
		t.Fatal(err)
	}
	if second.Previous != first.Release {
		t.Errorf("вторая выкатка не помнит прежний выпуск: %+v", second)
	}

	live := filepath.Join(server, "current", "app", "version.txt")
	if data, _ := os.ReadFile(live); string(data) != "второй" {
		t.Fatalf("работает %q, ожидали второй выпуск", data)
	}

	// Откат без всякой поломки: просто вернуться на шаг назад.
	back, err := p3k.Rollback(ctx, cfg)
	if err != nil {
		t.Fatalf("откат не удался: %v", err)
	}
	if back.Release != first.Release {
		t.Errorf("откатились на %q, ожидали %q", back.Release, first.Release)
	}
	if data, _ := os.ReadFile(live); string(data) != "первый" {
		t.Errorf("после отката работает %q", data)
	}

	// Дальше откатываться некуда — и об этом надо сказать, а не молча ничего
	// не сделать.
	if _, err := p3k.Rollback(ctx, cfg); err == nil {
		t.Error("откат с самого раннего выпуска должен отказать")
	}
}

func TestShipDryRunChangesNothing(t *testing.T) {
	project := t.TempDir()
	server := t.TempDir()
	os.Mkdir(filepath.Join(project, "app"), 0o755)
	os.WriteFile(filepath.Join(project, "app", "version.txt"), []byte("годная"), 0o644)

	config := fmt.Sprintf(`{"deploy": {"path": %q, "upload": ["app"], "checks": false}}`, server)
	os.WriteFile(filepath.Join(project, "p3k.json"), []byte(config), 0o644)
	cfg, err := p3k.Load(filepath.Join(project, "p3k.json"))
	if err != nil {
		t.Fatal(err)
	}

	var log strings.Builder
	if _, err := p3k.Ship(context.Background(), cfg, p3k.DryRun(), p3k.WithShipOutput(&log)); err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(log.String(), "Ничего не изменено") {
		t.Errorf("в отчёте нет главного: %s", log.String())
	}

	entries, _ := os.ReadDir(server)
	if len(entries) != 0 {
		t.Errorf("холостой прогон наследил на цели: %v", entries)
	}
}

func TestShipRefusesWithoutDeploySection(t *testing.T) {
	project := t.TempDir()
	os.WriteFile(filepath.Join(project, "p3k.json"), []byte(`{"processes":{"app":{"command":"echo"}}}`), 0o644)
	cfg, err := p3k.Load(filepath.Join(project, "p3k.json"))
	if err != nil {
		t.Fatal(err)
	}

	_, err = p3k.Ship(context.Background(), cfg)
	if err == nil || !strings.Contains(err.Error(), "нет секции deploy") {
		t.Errorf("ожидалось внятное объяснение, получили: %v", err)
	}
}

func TestUniqueReleaseKeepsOrder(t *testing.T) {
	base := p3k.ReleaseID(time.Date(2026, 9, 7, 17, 22, 56, 0, time.UTC))
	if base != "20260907-172256" {
		t.Fatalf("имя выпуска %q", base)
	}

	first, _ := p3k.UniqueRelease(nil, base)
	second, _ := p3k.UniqueRelease([]string{first}, base)
	third, _ := p3k.UniqueRelease([]string{first, second}, base)

	if !(first < second && second < third) {
		t.Errorf("имена перестали сортироваться по времени: %q %q %q", first, second, third)
	}
	if _, err := p3k.UniqueRelease([]string{}, base); err != nil {
		t.Errorf("свободное имя не должно давать ошибку: %v", err)
	}
}
